File: src/core/theme.zig
const std = @import("std");
const tui = @import("tui");

pub const Color = tui.Color;

pub const Theme = struct {
    background: Color,
    foreground: Color,
    surface: Color,
    surface_alt: Color,
    primary: Color,
    primary_bright: Color,
    on_primary: Color,
    on_primary_bright: Color,
    success: Color,
    success_bright: Color,
    danger: Color,
    danger_bright: Color,
    warning: Color,
    info: Color,
    info_bright: Color,
    accent: Color,
    accent_bright: Color,
    emphasis: Color,
    muted: Color,
    forge_github: Color,
    forge_gitlab: Color,
    forge_codeberg: Color,

    pub fn withTransparentBackground(self: Theme, enabled: bool) Theme {
        var theme = self;
        if (enabled) {
            theme.background = .default;
            theme.surface_alt = .default;
        }
        return theme;
    }

    pub fn lookup(self: Theme, key: []const u8) ?Color {
        inline for (comptime std.meta.fieldNames(Theme)) |field_name| {
            if (std.mem.eql(u8, key, field_name)) {
                return @field(self, field_name);
            }
        }
        return null;
    }
};

File: src/core/theme.rs
use crate::core::Color;

#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
pub struct Theme {
    pub background: Color,
    pub foreground: Color,
    pub surface: Color,
    pub surface_alt: Color,
    pub primary: Color,
    pub primary_bright: Color,
    pub on_primary: Color,
    pub on_primary_bright: Color,
    pub success: Color,
    pub success_bright: Color,
    pub danger: Color,
    pub danger_bright: Color,
    pub warning: Color,
    pub info: Color,
    pub info_bright: Color,
    pub accent: Color,
    pub accent_bright: Color,
    pub emphasis: Color,
    pub muted: Color,
    pub forge_github: Color,
    pub forge_gitlab: Color,
    pub forge_codeberg: Color,
}

impl Theme {
    pub const fn with_transparent_background(self, enabled: bool) -> Self {
        if enabled {
            Self {
                background: Color::Default,
                surface_alt: Color::Default,
                ..self
            }
        } else {
            self
        }
    }

    pub const fn lookup(self, key: &str) -> Option<Color> {
        match key.as_bytes() {
            b"background" => Some(self.background),
            b"foreground" => Some(self.foreground),
            b"surface" => Some(self.surface),
            b"surface_alt" => Some(self.surface_alt),
            b"primary" => Some(self.primary),
            b"primary_bright" => Some(self.primary_bright),
            b"on_primary" => Some(self.on_primary),
            b"on_primary_bright" => Some(self.on_primary_bright),
            b"success" => Some(self.success),
            b"success_bright" => Some(self.success_bright),
            b"danger" => Some(self.danger),
            b"danger_bright" => Some(self.danger_bright),
            b"warning" => Some(self.warning),
            b"info" => Some(self.info),
            b"info_bright" => Some(self.info_bright),
            b"accent" => Some(self.accent),
            b"accent_bright" => Some(self.accent_bright),
            b"emphasis" => Some(self.emphasis),
            b"muted" => Some(self.muted),
            b"forge_github" => Some(self.forge_github),
            b"forge_gitlab" => Some(self.forge_gitlab),
            b"forge_codeberg" => Some(self.forge_codeberg),
            _ => None,
        }
    }
}
